bot/utils/redis_cache.py

```diff
@@ -0,0 +1,373 @@
+        return items
+
+    async def length(self) -> int:
+        """Return the number of items in the Redis cache."""
+        await self._validate_cache()
+        number_of_items = await self._redis.hlen(self._namespace)
+        log.trace(f"Returning length. Result is {number_of_items}.")
+        return number_of_items
+
+    async def to_dict(self) -> Dict:
+        """Convert to dict and return."""
+        return {key: value for key, value in await self.items()}
+
+    async def clear(self) -> None:
+        """Deletes the entire hash from the Redis cache."""
+        await self._validate_cache()
+        log.trace("Clearing the cache of all key/value pairs.")
+        await self._redis.delete(self._namespace)
+
+    async def pop(self, key: RedisType, default: Optional[RedisType] = None) -> RedisType:
+        """Get the item, remove it from the cache, and provide a default if not found."""
+        log.trace(f"Attempting to pop {key}.")
+        value = await self.get(key, default)
+
+        log.trace(
+            f"Attempting to delete item with key '{key}' from the cache. "
+            "If this key doesn't exist, nothing will happen."
+        )
+        await self.delete(key)
+
+        return value
+
+    async def update(self, items: Dict[RedisType, RedisType]) -> None:
+        """
+        Update the Redis cache with multiple values.
+
+        This works exactly like dict.update from a normal dictionary. You pass
+        a dictionary with one or more key/value pairs into this method. If the keys
+        do not exist in the RedisCache, they are created. If they do exist, the values
+        are updated with the new ones from `items`.
+
+        Please note that both the keys and the values in the `items` dictionary
+        must consist of valid RedisTypes - ints, floats, or strings.
+        """
+        await self._validate_cache()
+        log.trace(f"Updating the cache with the following items:\n{items}")
+        await self._redis.hmset_dict(self._namespace, self._dict_to_typestring(items))
+
+    async def increment(self, key: RedisType, amount: Optional[int, float] = 1) -> None:
+        """
+        Increment the value by `amount`.
+
+        This works for both floats and ints, but will raise a TypeError
+        if you try to do it for any other type of value.
+
+        This also supports negative amounts, although it would provide better
+        readability to use .decrement() for that.
+        """
+        log.trace(f"Attempting to increment/decrement the value with the key {key} by {amount}.")
+
+        # Since this has several API calls, we need a lock to prevent race conditions
+        async with self._increment_lock:
+            value = await self.get(key)
+
+            # Can't increment a non-existing value
+            if value is None:
+                error_message = "The provided key does not exist!"
+                log.error(error_message)
+                raise KeyError(error_message)
```

Comment:
Without knowing how exactly the Redis Cache will be used, this may be a bit hard to predict, but do you have a general estimate of how frequently the caller might attempt to increment a key that doesn't exist? If it's expected to be reasonably often, it may be worth releasing the lock immediately after the `if value is None:` check.

However, if this is more for API safety and it's expected that the caller will be using a key that definitely exists 99% of the time, it's probably fine as it is.

Reply:
> do you have a general estimate of how frequently the caller might attempt to increment a key that doesn't exist?

There is no reason this should ever happen outside of the local environment. If someone wrote a feature that did this, we would probably catch it during review and prevent it from ever being merged. This feature won't be directly exposed to users (at least, I really don't think it should be), so in my opinion this should practically _never_ happen.